### PR TITLE
ObjectStore: Write json as json when possible

### PR DESCRIPTION
**Pull request by ryantxu**
This PR does three things aiming to make https://github.com/grafana/grafana/pull/56348 smaller
1. Adds an `objectStore` feature flag
2. Puts back custom JSON handling so that raw json bytes are written directly as JSON.  When it is not json, it is written to "body_base64"
3. Moves the "sync*" fields to their own protobuf object

image

### CODE REVIEW

File: pkg/services/store/object/dummy/dummy_server.go [hunk 1]
```diff
@@ -23,13 +24,14 @@ type ObjectVersionWithBody struct {
 }
 
 type RawObjectWithHistory struct {
-	*object.RawObject `json:"rawObject,omitempty"`
```
Comment (ryantxu) [inline pkg/services/store/object/dummy/dummy_server.go:26]: The embedded structs get really gnarly with JSON encoding, so I just switched to a normal one :_

File: pkg/services/store/object/object.proto [hunk 2]
```diff
@@ -286,10 +287,10 @@ message ObjectSearchResult {
   map<string,string> labels = 9;
 
   // Optionally include extracted JSON
-  string fields_json = 10;
+  bytes fields_json = 10;
```
Comment (ryantxu) [inline pkg/services/store/object/object.proto:290]: all other `_json` fields are bytes, so I think this is better

File: pkg/services/store/object/dummy/dummy_server.go [hunk 3]
```diff
@@ -109,10 +111,21 @@ func (i dummyObjectServer) Read(ctx context.Context, r *object.ReadObjectRequest
...
 
-	return &object.ReadObjectResponse{
-		Object:      objVersion,
-		SummaryJson: nil,
-	}, nil
+	rsp := &object.ReadObjectResponse{
+		Object: objVersion,
+	}
+	if r.WithSummary {
+		summary, _, e2 := object.GetSafeSaveObject(&object.WriteObjectRequest{
```
Comment (ArturWierzbicki) [inline pkg/services/store/object/dummy/dummy_server.go:118]: ha - this is interesting, so if its not the latest version we will have to recompute the summary on the fly

File: pkg/services/store/object/json.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+		stream.WriteString(obj.UpdatedBy)
+	}
+	if obj.Body != nil {
+		stream.WriteMore()
+		if json.Valid(obj.Body) {
+			stream.WriteObjectField("body")
+			stream.WriteRaw(string(obj.Body)) // works for strings
+		} else {
+			sEnc := base64.StdEncoding.EncodeToString(obj.Body)
+			stream.WriteObjectField("body_base64")
```
Comment (ArturWierzbicki) [inline pkg/services/store/object/json.go:95]: hmm - I think that's going to make swagger doc generation more difficult. what about keeping `body` for both and adding a `body_type` prop?

File: pkg/services/store/object/kind.go [hunk 5]
```diff
@@ -8,3 +13,47 @@ const StandardKindFolder = "folder"
 const StandardKindPanel = "panel"         // types: heatmap, timeseries, table, ...
 const StandardKindDataSource = "ds"       // types: influx, prometheus, test, ...
 const StandardKindTransform = "transform" // types: joinByField, pivot, organizeFields, ...
+
+// This is a stub -- it should lookup in a registry
```
Comment (ArturWierzbicki) [inline pkg/services/store/object/kind.go:17]: lets add a comment what is it supposed to do?

**Review by ArturWierzbicki — APPROVED** [review]
looks great overall - have just one concern about the `body_base64` prop

File: pkg/services/store/object/dummy/dummy_server.go (hunk 3 above)
Comment (ryantxu) [reply pkg/services/store/object/dummy/dummy_server.go:118]: in the dummy implementation... we never actually save the summary (we could)

Also note that by definition, the summary should always be able to be calculated from the raw bytes.  If they are out of sync then something went wrong and the new calculation is correct.

File: pkg/services/store/object/json.go (hunk 4 above)
Comment (ryantxu) [reply pkg/services/store/object/json.go:95]: It seems more clear to me that we define 'body' and 'body_base64' and say that one of them will be filled up -- but totally open to suggestions!  and obviously trying to use it for real will make the choice more clear

**Pull request merged**